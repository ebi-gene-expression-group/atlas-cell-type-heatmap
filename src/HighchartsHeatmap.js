import React from "react"
import PropTypes from 'prop-types'

import Highcharts from "highcharts"
import HighchartsReact from "highcharts-react-official"
import HC_heatmap from "highcharts/modules/heatmap"

// init the module
HC_heatmap(Highcharts)

const HighchartsHeatmap = props => {
  let matrixData = []
  let i=0, j=0
  if(props.axisData.x && props.heatmapData) {
    while(i < props.axisData.x.length){
      while(j < props.axisData.y.length){
        matrixData.push([i, j, Math.floor(props.heatmapData[j][i] * 1000) / 1000 + 1])
        j++
      }
      i++
      j=0
    }
  }

  const options = {
    chart: {
      type: `heatmap`,
      zoomType: `y`,
      height: props.axisData.y && props.axisData.y.length*20,
      width: props.axisData.x && props.axisData.x.length*100

    },
    credits: {
      enabled: false
    },
    labels: {
      step: 1
    },
    title: {
      text: `Heatmap for cell types vs marker genes`
    },

    xAxis: {
      categories: props.axisData.x,
      opposite: true,
      labels: {
        autoRotation: [-20]}
    },

    yAxis: {
      categories: props.axisData.y,
      credits: {
        enabled: false
      },
      title: false,
      reversed: true,
      startOnTick: false,
      endOnTick: false,
      gridLineWidth: 0,
      minorGridLineWidth: 0,
      showEmpty: false
    },

    tooltip: {
      formatter: function () {
        if(this.point.value === null) {
          return `<b>Cell Type:</b> ${this.series.xAxis.categories[this.point.x]} <br/> 
                  <b>Gene ID:</b> ${this.series.yAxis.categories[this.point.y]} <br /> 
                  <b>Average expression:</b> Not expressed <br/>`
        }
        else {
          return `<b>Cell Type:</b> ${this.series.xAxis.categories[this.point.x]} <br/> 
                  <b>Gene ID:</b> ${this.series.yAxis.categories[this.point.y]} <br /> 
                  <b>Average expression:</b> ${Math.round(this.point.value-1)} CPM <br/>`
        }
      }
    },

    colorAxis: {
      type: `logarithmic`,
      minColor: `#FFFFFF`,
      maxColor: `#024990`,
      stops: [[0, `#FFFFFF`], [0.67, `#9ecae1`], [1, `#3182bd`]],
    },

    legend: {
      align: `center`,
      verticalAlign: `top`,
      layout: `horizontal`,
      symbolWidth: 280
    },

    series: [
      {
        data: matrixData,
        nullColor: `#eaeaea`,
        states: {
          hover: {
            brightness: 0,
            borderColor: `coral`,
            borderWidth: 2
          }
        },
        borderWidth: 1,
        borderColor: `#dddddd`
      }
    ]
  }

  return <HighchartsReact highcharts={Highcharts} options={options} />
}

HighchartsHeatmap.propTypes = {
  axisData: PropTypes.shape({
    y: PropTypes.array,
    x: PropTypes.array
  }).isRequired,
  heatmapData: PropTypes.array.isRequired
}

export default HighchartsHeatmap
