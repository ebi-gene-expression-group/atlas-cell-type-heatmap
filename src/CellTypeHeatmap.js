import React from "react"
import PropTypes from 'prop-types'

import URI from 'urijs'
import _ from "lodash"

import HighchartsHeatmap from "./HighchartsHeatmap"
import CalloutAlert from './CalloutAlert'
import LoadingOverlay from './LoadingOverlay'
import PlotSettingsDropdown from './PlotSettingsDropdown'


class CellTypeHeatmap extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      expressionByCellTypeData: [],
      solrResponse: [],
      experimentAccessions: [],
      responseByAccession: [],
      axisData: {},
      selectedExperiment: null
    }
  }

  async _fetchExpressionDataForHeatmap({resource, host}) {
    this.setState({
      isLoading: true
    })

    //  `http://localhost:8080/gxa/sc/json/experiments/celltype/name/organism_part/value/pancreas`
    // "https://gist.githubusercontent.com/lingyun1010/095f414db81d34cef2a3bc8eaf56544f/raw/e898dd2ce08fb427e929df6c8dac262920f244a7/cellTypeAPIJsonResponse.json"

    const url = URI(resource, host).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      const results = Object.values(await response.json())[0]
      const experimentAccessions = Object.keys(results)
      const markerGenes = Object.keys(results[experimentAccessions[0]]).sort()

      this.setState({
        markerGenes: markerGenes,
        experimentAccessions: Object.keys(results),
        expressionByCellTypeData: response,
        responseByAccession: results,
        axisData: this._heatmapAxis(markerGenes, results),
        heatmapData: this._factorHeatmapData(markerGenes, results),
        filteredData: this._factorHeatmapData(markerGenes, results),
        isLoading: false,
        selectedExperiment: null
      })

    } catch(e) {
      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  _factorHeatmapData(markerGenes, results, selectedExperiment) {
    const experimentAccessions = selectedExperiment ? [selectedExperiment] : Object.keys(results)
    const cellTypes = experimentAccessions.map(accession =>
      Object.keys(results[accession][Object.keys(results[accession])[0]])
    )
    const heatmapData = markerGenes.map(markerGene =>
      _.flatMap(
        experimentAccessions.map((experimentAccession, idx) =>
          cellTypes[idx].map(cellType => {
            return results[experimentAccession][markerGene][cellType]
          })
        )
      )
    )
    return heatmapData
  }

  _heatmapAxis(markerGenes, results, selectedExperiment) {
    const experimentAccessions = selectedExperiment ? [selectedExperiment] : Object.keys(results)
    const cellTypes = experimentAccessions.map(accession =>
      Object.keys(results[accession][Object.keys(results[accession])[0]])
    )

    const yAxisCategory = _.flatMap(
      experimentAccessions.map((experimentAccession, idx) =>
        cellTypes[idx].map(
          cellType => `${cellType} <br> ${experimentAccession}`
        )
      )
    )
    return ({ y: markerGenes, x: yAxisCategory })
  }

  componentDidMount() {
    this._fetchExpressionDataForHeatmap(this.props)
  }

  componentDidUpdate(previousProps) {
    if (previousProps.resource !== this.props.resource) {
      this._fetchExpressionDataForHeatmap(this.props)
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

  render() {
    const { heatmapData, axisData, hasError, isLoading } = this.state
    const { wrapperClassName, plotWrapperClassName } = this.props
    const { hasDynamicHeight, defaultHeatmapHeight, heatmapRowHeight } = this.props
    const { selectedExperiment, experimentAccessions, markerGenes, responseByAccession, filteredData } = this.state

    const experimentOptions = experimentAccessions
      .sort((a, b) => a-b)
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    if (experimentOptions.length > 1) {
      experimentOptions.unshift({
        value: `all`,
        label: `All`,
        isDisabled: false
      })
    }

    return (
      hasError ?
        <CalloutAlert error={hasError}/> :
        <div>
          <div className={wrapperClassName}>
            <div className={`small-12 medium-6 columns`}>
              <PlotSettingsDropdown
                labelText={`Show Gene Expressions for Experiment:`}
                options={experimentOptions}
                onSelect={(selectedOption) =>{
                  this.setState({
                    filteredData: selectedOption.value === `all` ?
                      _.cloneDeep(heatmapData) :
                      this._factorHeatmapData(markerGenes, responseByAccession, selectedOption.value),
                    axisData: selectedOption.value === `all` ?
                      this._heatmapAxis(markerGenes, responseByAccession) :
                      this._heatmapAxis(markerGenes, responseByAccession, selectedOption.value),
                    selectedExperiment: selectedOption
                  })
                }}
                value={selectedExperiment || experimentOptions[0]}
              />
            </div>
          </div>
          <div className={wrapperClassName}>
            <div className={plotWrapperClassName} style={{position: `relative`}}>
              <HighchartsHeatmap
                axisData={axisData}
                heatmapData={filteredData}
                chartHeight={defaultHeatmapHeight}
                hasDynamicHeight={axisData.y && axisData.y.length > 5 ? hasDynamicHeight : false} // don't want dynamic height if there is little or no data
                heatmapRowHeight={heatmapRowHeight}
              />
              <LoadingOverlay
                show={isLoading}
              />
            </div>
          </div>
        </div>
    )
  }
}

CellTypeHeatmap.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,
  plotWrapperClassName: PropTypes.string,
  defaultHeatmapHeight: PropTypes.number,
  hasDynamicHeight: PropTypes.bool,
  heatmapRowHeight: PropTypes.number
}

CellTypeHeatmap.defaultProps = {
  wrapperClassName: `row`,
  plotWrapperClassName: `small-12 columns`,
  defaultHeatmapHeight: 300,
  hasDynamicHeight: true,
  heatmapRowHeight: 20
}
export default CellTypeHeatmap
