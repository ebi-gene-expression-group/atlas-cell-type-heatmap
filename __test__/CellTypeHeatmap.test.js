import React from 'react'
import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'

import '@babel/polyfill'
import fetchMock from 'fetch-mock'

import CalloutAlert from '../src/CalloutAlert'
import CellTypeHeatmap from '../src/CellTypeHeatmap'
import HighchartsHeatmap from '../src/HighchartsHeatmap'

Enzyme.configure({ adapter: new Adapter() })

describe(`CellTypeHeatmap`, () => {
  beforeEach(() => {
    fetchMock.restore()
  })

  const props = {
    host: `foo/`,
    resource: `bar`
  }

  const validProps = {
    host: 'http://localhost:8080/gxa/sc/',
    resource: 'json/experiments/celltype/name/organism_part/value/pancreas'
  }

  test(`Renders error if API request is unsuccessful`, () => {
    const wrapper = shallow(<CellTypeHeatmap {...props} />)
    expect(wrapper.exists(CalloutAlert)).toBe(true)
  })

  test(`Renders heatmap if API request is successful`, () => {
    const wrapper = mount(<CellTypeHeatmap {...validProps} />)

    Promise
      .resolve(wrapper)
      .then(() => wrapper.update())
      .then(() => wrapper.update())
      .then(() => {
        expect(wrapper.exists(HighchartsHeatmap)).toBe(true)
      })
  })

})
