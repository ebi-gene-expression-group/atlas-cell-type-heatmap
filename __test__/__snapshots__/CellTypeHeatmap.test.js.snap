// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellTypeHeatmap matches snapshot 1`] = `
<div
  className="row"
>
  <div
    className="columns large-9 large-centered"
  >
    <div
      className="callout alert small"
    >
      <h5>
        Oops!
      </h5>
      <p>
        There was a problem rendering this component.
        <br />
        If the error persists, in order to help us debug the issue, please copy the URL and this message and send it to us via 
        <a
          href="https://www.ebi.ac.uk/support/gxasc"
        >
          the EBI Support & Feedback system
        </a>
        :
      </p>
      <code>
        TypeError: Cannot read property 'nodeName' of null â€“ [object Object]
      </code>
    </div>
  </div>
</div>
`;
